### PR TITLE
✨ Source-Stripe: Add new payout_balance_transactions incremental stream

### PR DESCRIPTION
**Pull request by williamkaper**
## What
This PR adds a new stream that shows what balance transactions made up an automatic payout ([Stripe API Link - called with payout ID as parameter](https://docs.stripe.com/api/balance_transactions/list)). This solves long standing [issue/request #8125](https://github.com/airbytehq/airbyte/issues/8125). So if a company collects 25 charges and then pays them out using Stripe's automatic payout functionality, this new stream would show the linkage between the 25 charge balance transactions and the payout. 

## How
I used the pattern laid by other child streams that make piggy back calls for a parent against a rest endpoint for embedded entires ike checkout_session_line_items. This is a little different in that the endpoint is not called with the ID of the parent in the URI, but instead the balance transaction API is called with a parameter of `payout='po_id'`. I use the payout entity as the parent, and use its ID to call the API and store the payout ID along with the balance transaction records associated to the payout, and the updated timestamp of the payout record so the new stream can also handle incremental updates. 

## Review guide

1. source.py - added code for new sub stream
2. erd/discovered_catalog.json
3. erd/source.dbml (Note - Please check how I did the composite Primary key to make sure its right)
4. source_stripe/schemas/payout_balance_transactions.json
5. unit_tests/test_source.py -- Changed unit test to expect one additional stream.  

## User Impact
This new stream will only be required for users who want to see the linkage a payout and the balance transactions that comprised it. Manually initiated payouts will NOT have any linkage to charge balance transactions or fee balance transactions. 

## Can this PR be safely reverted and rolled back?
- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json
Comment (williamkaper) [inline airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json:4010]: Not sure if I need this or just the source_defined_primary_key - I saw conflicting documentation on if airbyte did or did now support composite primary keys. In theory ID would be enough, but I don't know 100% if a balance transaction could be linked to multiple payouts.

Comment (natikgadzhi) [reply airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json]: composite PKs made of two+ columns generally don't work very well in Airbyte, and I remember we're trying to avoid them.

File: airbyte-integrations/connectors/source-stripe/pyproject.toml
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-stripe/pyproject.toml:6]: bump to `5.7.0`

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json]: Maybe in this case we can create an artificial PK combining `{id}_{payout}` @maxi297 do you have any thoughts here?

Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/erd/discovered_catalog.json]: We can probably do that, but can @natikgadzhi expand on this comment? I thought Airbyte wasn't dropping support for nested PKs but not composite ones. I would like to clarify that because we have a lot of sources that have composite PKs like source-google-analytics-data-api or source-google-search-console

File: airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json:59]: I don't see this field in the documentation or in the return value. Where does this come from?

**Review by maxi297 — COMMENTED** [review]
Awesome! Thanks for your contribution @williamkaper ! 

Can we add documentation on this new stream? It should be listed in stripe.md. We should also document why we are adding `payout_balance_transactions` and not just `balance_transactions`. My understanding is that the foreign key `payout` in `balance_transactions` is not always present in the responses from Stripe. Here is an example in our sandbox account where the balance transactions endpoint returns three entries for payout `po_1KVQhfEcXtiJtvvhZlUkl08U` although only the first one explicitly has `po_1KVQhfEcXtiJtvvhZlUkl08U` in the record:

```
~% curl -G 'https://api.stripe.com/v1/balance_transactions?payout=po_1KVQhfEcXtiJtvvhZlUkl08U' -u "<key>:"
{
  "object": "list",
  "data": [
    {
      "id": "txn_1KVQhfEcXtiJtvvhF7ox3YEm",
      "object": "balance_transaction",
      "amount": -9164,
      "available_on": 1645488000,
      "created": 1645406919,
      "currency": "usd",
      "description": "STRIPE PAYOUT",
      "exchange_rate": null,
      "fee": 0,
      "fee_details": [],
      "net": -9164,
      "reporting_category": "payout",
      "source": "po_1KVQhfEcXtiJtvvhZlUkl08U",
      "status": "available",
      "type": "payout"
    },
    {
      "id": "txn_3K9FSOEcXtiJtvvh0KoS5mx7",
      "object": "balance_transaction",
      "amount": 5300,
      "available_on": 1640649600,
      "created": 1640120473,
      "currency": "usd",
      "description": null,
      "exchange_rate": null,
      "fee": 184,
      "fee_details": [
        {
          "amount": 184,
          "application": null,
          "currency": "usd",
          "description": "Stripe processing fees",
          "type": "stripe_fee"
        }
      ],
      "net": 5116,
      "reporting_category": "charge",
      "source": "ch_3K9FSOEcXtiJtvvh0zxb7clc",
      "status": "available",
      "type": "charge"
    },
    {
      "id": "txn_3K9F5DEcXtiJtvvh1qsqmHcH",
      "object": "balance_transaction",
      "amount": 4200,
      "available_on": 1640649600,
      "created": 1640119035,
      "currency": "usd",
      "description": "edgao test",
      "exchange_rate": null,
      "fee": 152,
      "fee_details": [
        {
          "amount": 152,
          "application": null,
          "currency": "usd",
          "description": "Stripe processing fees",
          "type": "stripe_fee"
        }
      ],
      "net": 4048,
      "reporting_category": "charge",
      "source": "ch_3K9F5DEcXtiJtvvh1w2MaTpj",
      "status": "available",
      "type": "charge"
    }
  ],
  "has_more": false,
  "url": "/v1/balance_transactions"
}%
```

Last thing is: I've added some tests in order to understand the behavior. You can see those tests on [this commit](https://github.com/airbytehq/airbyte/commit/afacf6a2b909dc61079837334f749d91db1d62ee). It would be great to keep those tests and if you see any other test cases to be added, feel free to do so.

**Comment by williamkaper** [comment]
Hi @maxi297 , I want to try to get all your TODOs/questions answered today to help unblock this. First off - thank you for adding the additional tests - that is super helpful to help see how those work.

> Can we add documentation on this new stream? 
Yes I will do that now. 

> We should also document why we are adding payout_balance_transactions and not just balance_transactions. My understanding is that the foreign key payout in balance_transactions is not always present in the responses from Stripe. Here is an example in our sandbox account where the balance transactions endpoint returns three entries for payout po_1KVQhfEcXtiJtvvhZlUkl08U although only the first one explicitly has po_1KVQhfEcXtiJtvvhZlUkl08U in the record:
Great question. There are two concepts here that are confusing. There is an actual balance transaction for a payout (think of a checkbook ledger entry when you do a withdrawl of $1000 from the bank). That is what you are seeing the source linking to. What that doesn't tell you is what deposits went into your bank account that contributed to the $1000, which is why we need to make a special API call with the payout ID as a parameter to see what balance transactions actually comprised the $1000 payout. One of the transactions it returns is the transaction that withdrawls the $1000. Honestly, Stripe's engineers seem to enjoy being inconsistent with how they do stuff.

File: airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json
Comment (williamkaper) [reply airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json:59]: @maxi297 I don't know. I copied `airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/schemas/shared/balance_transactions.json` to create the payout_balance_transaction json and just added `payout` and `updated` to it and didn't think to check anything else since payout balance transaction is really a grouped copy of balance transaction grouped on payout.

Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json:59]: Ok! [This](https://docs.stripe.com/upgrades#2017-01-27) probably answers the question. It was probably kept in `balance_transactions.json` because of legacy reasons but would remove it for the new stream.

Comment (williamkaper) [reply airbyte-integrations/connectors/source-stripe/source_stripe/schemas/payout_balance_transactions.json:59]: @maxi297 Ok I'll remove it now

**Comment by williamkaper** [comment]
I also changed the version number to 5.8.1 and merge updated with master to bring the entire branch up to date.

**Review by maxi297 — COMMENTED** [review]
Can we merge the tests from https://github.com/airbytehq/airbyte/commit/afacf6a2b909dc61079837334f749d91db1d62ee? After this, I'm good to approve

**Comment by williamkaper** [comment]
Merged.

**Review by maxi297 — APPROVED** [review]
Awesome work! Thanks @williamkaper for contributing to Airbyte and addressing the comments on this PR 💜

**Comment by williamkaper** [comment]
> Awesome work! Thanks @williamkaper for contributing to Airbyte and addressing the comments on this PR 💜

Happy to do it. If there is still any issues that are minor, would you mind cleaning them up for me?

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**
